pkg/derivation/encode.go

```diff
@@ -77,12 +78,27 @@ func encodeArray(opening byte, closing byte, quote bool, elems ...[]byte) []byte
 	return buf.Bytes()
 }
 
-// WriteDerivation writes the textual representation of the derivation to the passed writer.
-func (d *Derivation) WriteDerivation(writer io.Writer) error {
-	// we need to sort outputs by their name, which is the key of the map.
-	// get the list of keys, sort them, then add each one by one.
-	// Due to the "sorted paths" requirement, we know there's no two
-	// outputs with the same path.
+// WriteDerivation writes the ATerm representation of the derivation to the passed writer.
+// Optionally, the following transformations can be made while writing out the ATerm:
+//
+// - stripOutput will replace output hashes in `Outputs` (`Output[$outputName]`),
+//   and `env[$outputName]` with empty strings
+//
+// - inputDrvReplacements (map[$drvPath]$replacement) can be provided.
+//   If set, it must contain all derivation path in d.InputDerivations[*]
+//   These will be replaced with their replacement value.
+//   As this will change map keys, and map keys need to be serialized alphabetically sorted,
+//   this will shuffle the order of values.
+//
+// This replacement/stripping is only used when calculating output hashes.
+// Set to false / nil in normal mode.
```

Comment:
It seems like this is something we might want to hide. WriteDerivation would only get a io.Writer, but then call writeDerivation that takes the stripOutputs and inputDrvReplacements, with empty values

Reply:
Indeed, we don't use these modes outside code in `package derivation` anymore.

These modes are only used in `getMaskedATermHash` (which is invoked by `CalculateOutputPaths`), and `CalculateDrvReplacement`.